File: solutions/cpp/remove_duplicates_sorted_list.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Linked_list_prototype.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::shared_ptr;

// @include
shared_ptr<ListNode<int>> RemoveDuplicates(
    const shared_ptr<ListNode<int>>& L) {
  auto iter = L;
  while (iter) {
    // Uses next_distinct to find the next distinct value.
    auto next_distinct = iter->next;
    while (next_distinct && next_distinct->data == iter->data) {
      next_distinct = next_distinct->next;
    }
    iter->next = next_distinct;
    iter = next_distinct;
  }
  return L;
}
// @exclude

void SimpleTest() {
  shared_ptr<ListNode<int>> L = nullptr;
  auto result = RemoveDuplicates(L);
  assert(result == nullptr);
  L = make_shared<ListNode<int>>(ListNode<int>{123, nullptr});
  result = RemoveDuplicates(L);
  assert(result == L);
  L->next = make_shared<ListNode<int>>(ListNode<int>{123, nullptr});
  result = RemoveDuplicates(L);
  assert(result->next == nullptr);

  // Creating an invalid input, 123 -> 124 -> 123, algo will not detect dups!
  L->next = make_shared<ListNode<int>>(ListNode<int>{124, nullptr});
  L->next->next = make_shared<ListNode<int>>(ListNode<int>{123, nullptr});
  result = RemoveDuplicates(L);
  assert(result->data == 123 && result->next->data == 124 &&
         result->next->next->data == 123);
}

int main(int argc, char** argv) {
  SimpleTest();
  shared_ptr<ListNode<int>> L = make_shared<ListNode<int>>(ListNode<int>{
      2, make_shared<ListNode<int>>(ListNode<int>{
             2, make_shared<ListNode<int>>(ListNode<int>{
                    2, make_shared<ListNode<int>>(ListNode<int>{
                           2, make_shared<ListNode<int>>(
                                  ListNode<int>{2, nullptr})})})})});
  shared_ptr<ListNode<int>> pre = nullptr;
  auto result = RemoveDuplicates(L);
  int count = 0;
  while (result) {
    ++count;
    if (pre) {
      assert(pre->data != result->data);
    }
    cout << result->data << endl;
    pre = result;
    result = result->next;
  }
  assert(count == 1);
  return 0;
}

File: static/python/remove_duplicates_sorted_list.py
from linked_list_prototype import ListNode


# @include
def remove_duplicates(L):
    it = L
    while it:
        # Uses next_distinct to find the next distinct value.
        next_distinct = it.next
        while next_distinct and next_distinct.data == it.data:
            next_distinct = next_distinct.next
        it.next = next_distinct
        it = next_distinct
    return L
# @exclude


def simple_test():
    L = None
    assert not remove_duplicates(L)
    L = ListNode(123)
    result = remove_duplicates(L)
    assert result is L
    L.next = ListNode(123)
    result = remove_duplicates(L)
    assert not result.next

    # Creating an invalid input, 123 -> 124 -> 123, algo will not detect dups!
    L.next = ListNode(124, ListNode(123))
    result = remove_duplicates(L)
    assert L.data == 123 and L.next.data == 124 and L.next.next.data == 123


def main():
    simple_test()
    L = ListNode(2, ListNode(2, ListNode(2, ListNode(2, ListNode(2, None)))))
    pre = None
    result = remove_duplicates(L)
    count = 0
    while result:
        count += 1
        if pre:
            assert pre.data != result.data
        print(result.data)
        pre = result
        result = result.next
    assert count == 1


if __name__ == '__main__':
    main()
